File: news_server/modules/news/presentation/eventbus/handler/news.ts
/**
 * News 事件处理器
 * 类似 Sjgz-Backend 的 interfaces/eventbus/handler/news.go
 */
import { consola } from 'consola'
import type { NewsService } from '../../../application/news/service'
import type { SourceDataFetchedEvent } from '@events/source'

export class NewsHandler {
  constructor(private readonly newsService: NewsService) {}

  /**
   * 处理源数据已获取事件
   * 将 Source 服务获取的数据存储到数据库
   */
  async OnSourceDataFetched(eventData: SourceDataFetchedEvent): Promise<void> {
    consola.info('✅ [News Worker] 已收到源数据获取事件:', {
      sourceId: eventData.sourceId,
      itemsCount: eventData.items.length,
    })

    try {
      // 转换为 NewsCreateWO 格式
      const items = eventData.items.map(item => ({
        sourceId: eventData.sourceId,
        originalId: item.id,
        title: item.title,
        url: item.url,
        mobileUrl: item.mobileUrl,
        pubDate: item.pubDate,
        extra: item.extra,
      }))

      // 批量创建新闻
      await this.newsService.createNewsBatch.execute({ items })
      
      consola.success(`✅ [News Worker] 源 ${eventData.sourceId} 的 ${items.length} 条新闻已存储`)
    } catch (error) {
      consola.error(`❌ [News Worker] 处理源数据失败:`, error)
      throw error
    }
  }
}

File: modules/news/interfaces/grpc/handler/news.go
package handler

import (
	"context"
	"encoding/json"

	"nfxnews/events"
	newsapp "nfxnews/modules/news/application/news"
	newspb "nfxnews/protos/gen/news"
)

type NewsHandler struct {
	newspb.UnimplementedNewsServiceServer
	svc *newsapp.Service
}

func NewNewsHandler(svc *newsapp.Service) *NewsHandler { return &NewsHandler{svc: svc} }

func toPB(items []newsapp.ItemView) []*newspb.NewsItem {
	out := make([]*newspb.NewsItem, 0, len(items))
	for _, it := range items {
		extra, _ := json.Marshal(it.Extra)
		out = append(out, &newspb.NewsItem{
			Id: it.ID, SourceId: it.SourceID, OriginalId: it.OriginalID, Title: it.Title,
			Url: it.URL, MobileUrl: it.MobileURL, PubDateUnix: it.PubDate, ExtraJson: string(extra),
		})
	}
	return out
}

func (h *NewsHandler) UpsertItems(ctx context.Context, req *newspb.UpsertItemsRequest) (*newspb.UpsertItemsResponse, error) {
	ev := events.SourceFetchedEvent{SourceID: req.GetSourceId()}
	for _, it := range req.GetItems() {
		ev.Items = append(ev.Items, events.SourceNewsItem{
			ID: it.GetOriginalId(), Title: it.GetTitle(), URL: it.GetUrl(), MobileURL: it.GetMobileUrl(),
			PubDate: it.GetPubDateUnix(), ExtraJSON: it.GetExtraJson(),
		})
	}
	n, err := h.svc.UpsertFromEvent(ctx, ev)
	if err != nil {
		return nil, err
	}
	return &newspb.UpsertItemsResponse{Upserted: int32(n)}, nil
}

func (h *NewsHandler) ListBySource(ctx context.Context, req *newspb.ListBySourceRequest) (*newspb.ListBySourceResponse, error) {
	items, err := h.svc.ListBySource(ctx, req.GetSourceId(), int(req.GetLimit()))
	if err != nil {
		return nil, err
	}
	return &newspb.ListBySourceResponse{Items: toPB(items)}, nil
}

func (h *NewsHandler) SearchNews(ctx context.Context, req *newspb.SearchNewsRequest) (*newspb.SearchNewsResponse, error) {
	items, err := h.svc.Search(ctx, req.GetQuery(), int(req.GetLimit()))
	if err != nil {
		return nil, err
	}
	return &newspb.SearchNewsResponse{Items: toPB(items)}, nil
}
